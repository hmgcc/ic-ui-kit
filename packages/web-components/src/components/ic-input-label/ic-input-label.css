ic-input-label {
  margin-bottom: var(--ic-space-xs);

  --ic-typography-color: var(
    --ic-input-label-text-color,
    var(--ic-color-text-primary)
  );
}

ic-input-label.with-helper {
  margin-bottom: var(--ic-space-xxxs);
}

ic-input-label.ic-input-label-readonly {
  --ic-typography-color: var(--ic-input-label-readonly-typography-color);
}

ic-input-label .helpertext,
ic-input-label .helpertext ::slotted(*) {
  margin-top: var(--ic-space-xxxs);
  padding-bottom: var(--ic-input-label-helpertext-padding, 0);

  --ic-typography-color: var(
    --ic-input-label-helper-text-color,
    var(--ic-color-text-secondary)
  );
}

ic-input-label .helpertext-normal {
  --ic-typography-color: var(
    --ic-input-label-helper-text-color,
    var(--ic-color-text-secondary)
  );
}

ic-input-label .helpertext-readonly {
  --ic-typography-color: var(--ic-input-label-readonly-helpertext-color);
}

ic-input-label .readonly-label {
  --ic-typography-color: var(--ic-input-label-readonly-label-color);
}

ic-input-label .error-label {
  --ic-typography-color: var(--ic-input-label-error-label-color);
}
