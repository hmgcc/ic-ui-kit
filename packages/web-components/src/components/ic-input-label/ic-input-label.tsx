import { Component, Element, Host, Prop, h } from "@stencil/core";

import {
  getInputHelperTextID,
  onComponentRequiredPropUndefined,
} from "../../utils/helpers";

@Component({
  tag: "ic-input-label",
  styleUrl: "./ic-input-label.css",
})
export class InputLabel {
  @Element() el: HTMLIcInputLabelElement;

  /**
   * If `true`, the disabled state will be set.
   */
  @Prop() disabled?: boolean = false;

  /**
   * The ID of the form element the label is bound to.
   */
  @Prop() for?: string;

  /**
   * The helper text that will be displayed.
   */
  @Prop() helperText: string = "";

  /**
   * The label will be visually hidden.
   */
  @Prop() hideLabel: boolean = false;

  /**
   * The text content of the label.
   */
  @Prop() label!: string;

  /**
   * If `true`, the readonly state will be set.
   */
  @Prop() readonly: boolean = false;

  /**
   * If `true`, the input label will require a value.
   */
  @Prop() required: boolean = false;

  /**
   * The status of the label - e.g. 'error'.
   */
  @Prop() status: "error" | "" = "";

  /**
   * @internal If `true`, wraps label text in label tag
   */
  @Prop() useLabelTag: boolean = true;

  componentDidLoad(): void {
    onComponentRequiredPropUndefined(
      [{ prop: this.label, propName: "label" }],
      "Input Label"
    );
  }

  private isHelperTextSlotUsed = (slot: Element | null): boolean => {
    const assignedEls = (slot as HTMLSlotElement)?.assignedElements();
    if (assignedEls && assignedEls.length) {
      for (const el of assignedEls) {
        if (el.tagName === "SLOT") {
          // Recursion needed for when slot is forwarded multiple times - through child components
          // (e.g. in date picker)
          if (this.isHelperTextSlotUsed(el as HTMLSlotElement)) {
            return true;
          }
        } else {
          // Found an assigned element which is not a nested <slot>
          return true;
        }
      }
    }
    return false;
  };

  render() {
    const {
      disabled,
      readonly,
      label,
      required,
      helperText,
      status,
      hideLabel,
      useLabelTag,
    } = this;
    const labelText = required ? label + " *" : label;
    const helperTextId = this.for && getInputHelperTextID(this.for);
    const helperTextClass = {
      helpertext: true,
      "helpertext-normal": !disabled && !readonly,
      "helpertext-readonly": readonly,
    };

    const helperTextSlot = this.el.querySelector("slot[name='helper-text']");

    return (
      <Host
        class={{
          "ic-input-label-disabled": !!disabled,
          "ic-input-label-readonly": readonly,
          "with-helper":
            this.isHelperTextSlotUsed(helperTextSlot) || helperText !== "",
        }}
      >
        {!hideLabel && (
          <ic-typography
            variant="label"
            class={{
              "readonly-label": readonly,
              "error-label": status === "error" && !(readonly || disabled),
            }}
          >
            {readonly || !useLabelTag ? (
              `${labelText}`
            ) : (
              <label htmlFor={this.for}>{labelText}</label>
            )}
          </ic-typography>
        )}
        {this.isHelperTextSlotUsed(helperTextSlot) ? (
          <span id={helperTextId} class={helperTextClass}>
            <slot name="helper-text"></slot>
          </span>
        ) : (
          helperText !== "" && (
            <ic-typography variant="caption" class={helperTextClass}>
              <span id={helperTextId}>{helperText}</span>
            </ic-typography>
          )
        )}
      </Host>
    );
  }
}
