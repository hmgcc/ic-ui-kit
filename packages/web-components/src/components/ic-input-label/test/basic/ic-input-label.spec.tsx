import { InputLabel } from "../../ic-input-label";
import { newSpecPage } from "@stencil/core/testing";

describe("ic-input-label", () => {
  it("should render", async () => {
    const page = await newSpecPage({
      components: [InputLabel],
      html: `<ic-input-label for="test-input-id" label="Test label"></ic-input-label>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render required variant", async () => {
    const page = await newSpecPage({
      components: [InputLabel],
      html: `<ic-input-label for="test-input-id" label="Test label" required=true></ic-input-label>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render disabled variant", async () => {
    const page = await newSpecPage({
      components: [InputLabel],
      html: `<ic-input-label for="test-input-id" label="Test label" required=true disabled=true></ic-input-label>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render readonly variant", async () => {
    const page = await newSpecPage({
      components: [InputLabel],
      html: `<ic-input-label for="test-input-id" label="Test label" required=true readonly=true></ic-input-label>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render error variant", async () => {
    const page = await newSpecPage({
      components: [InputLabel],
      html: `<ic-input-label for="test-input-id" label="Test label" status="error"></ic-input-label>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with helpertext", async () => {
    const page = await newSpecPage({
      components: [InputLabel],
      html: `<ic-input-label for="test-input-id" label="Test label" required=true helper-text="Some helper text"></ic-input-label>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with label text not wrapped in label tag", async () => {
    const page = await newSpecPage({
      components: [InputLabel],
      html: `<ic-input-label label="Test label" required=true helper-text="Some helper text" use-label-tag=false></ic-input-label>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should remove label but keep helpertext", async () => {
    const page = await newSpecPage({
      components: [InputLabel],
      html: `<ic-input-label for="test-input-id" label="Test label" hide-label='true' helper-text="Some helper text"></ic-input-label>`,
    });

    expect(page.root).toEqualHtml(`
    <ic-input-label class="with-helper" for="test-input-id" helper-text="Some helper text" hide-label="true" label="Test label">
      <ic-typography class="helpertext helpertext-normal" variant="caption">
        <span id="test-input-id-helper-text">
          Some helper text
        </span>
      </ic-typography>
    </ic-input-label>
    `);
  });

  it("should remove helpertext but keep label", async () => {
    const page = await newSpecPage({
      components: [InputLabel],
      html: `<ic-input-label for="test-input-id" label="Test label" helper-text=""></ic-input-label>`,
    });

    expect(page.root).toEqualHtml(`
      <ic-input-label for="test-input-id" helper-text="" label="Test label">
        <ic-typography variant="label">
          <label htmlfor="test-input-id">
            Test label
          </label>
        </ic-typography>
      </ic-input-label>
    `);
  });

  it("should correctly detect if a helper text slot is used", async () => {
    const page = await newSpecPage({
      components: [InputLabel],
      html: `<ic-input-label for="test-input-id" label="Test label"></ic-input-label>`,
    });

    const mockSlot = {
      assignedElements: jest
        .fn()
        .mockReturnValue([document.createElement("div")]),
    } as unknown as HTMLSlotElement;

    expect(page.rootInstance.isHelperTextSlotUsed(mockSlot)).toBe(true);

    const parentMockSlot = {
      assignedElements: jest.fn().mockReturnValue([mockSlot]),
    } as unknown as HTMLSlotElement;

    parentMockSlot.assignedElements = jest.fn().mockReturnValue([mockSlot]);

    expect(page.rootInstance.isHelperTextSlotUsed(parentMockSlot)).toBe(true);

    mockSlot.assignedElements = jest.fn().mockReturnValue([]);

    expect(page.rootInstance.isHelperTextSlotUsed(mockSlot)).toBe(false);
  });
});
