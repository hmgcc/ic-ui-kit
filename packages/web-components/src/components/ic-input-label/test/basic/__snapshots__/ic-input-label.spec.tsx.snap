// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-input-label should render 1`] = `
<ic-input-label for="test-input-id" label="Test label">
  <!---->
  <ic-typography variant="label">
    <label htmlfor="test-input-id">
      Test label
    </label>
  </ic-typography>
</ic-input-label>
`;

exports[`ic-input-label should render disabled variant 1`] = `
<ic-input-label class="ic-input-label-disabled" disabled="true" for="test-input-id" label="Test label" required="true">
  <!---->
  <ic-typography variant="label">
    <label htmlfor="test-input-id">
      Test label *
    </label>
  </ic-typography>
</ic-input-label>
`;

exports[`ic-input-label should render error variant 1`] = `
<ic-input-label for="test-input-id" label="Test label" status="error">
  <!---->
  <ic-typography class="error-label" variant="label">
    <label htmlfor="test-input-id">
      Test label
    </label>
  </ic-typography>
</ic-input-label>
`;

exports[`ic-input-label should render readonly variant 1`] = `
<ic-input-label class="ic-input-label-readonly" for="test-input-id" label="Test label" readonly="true" required="true">
  <!---->
  <ic-typography class="readonly-label" variant="label">
    Test label *
  </ic-typography>
</ic-input-label>
`;

exports[`ic-input-label should render required variant 1`] = `
<ic-input-label for="test-input-id" label="Test label" required="true">
  <!---->
  <ic-typography variant="label">
    <label htmlfor="test-input-id">
      Test label *
    </label>
  </ic-typography>
</ic-input-label>
`;

exports[`ic-input-label should render with helpertext 1`] = `
<ic-input-label class="with-helper" for="test-input-id" helper-text="Some helper text" label="Test label" required="true">
  <!---->
  <ic-typography variant="label">
    <label htmlfor="test-input-id">
      Test label *
    </label>
  </ic-typography>
  <ic-typography class="helpertext helpertext-normal" variant="caption">
    <span id="test-input-id-helper-text">
      Some helper text
    </span>
  </ic-typography>
</ic-input-label>
`;

exports[`ic-input-label should render with label text not wrapped in label tag 1`] = `
<ic-input-label class="with-helper" helper-text="Some helper text" label="Test label" required="true" use-label-tag="false">
  <!---->
  <ic-typography variant="label">
    Test label *
  </ic-typography>
  <ic-typography class="helpertext helpertext-normal" variant="caption">
    <span>
      Some helper text
    </span>
  </ic-typography>
</ic-input-label>
`;
